### ISSUE TITLE
Nested function accessing parent state

### ISSUE THREAD

**Issue by tindzk** [scala-native/scala-native#707]
The following example does not compile:

```scala
  def test(f: CFunctionPtr1[CInt, Unit]): Unit = extern
  def f(i: Int): Unit = {
    def g(j: CInt): Unit = println(i + j)
    test(CFunctionPtr.fromFunction1(g))
  }
```

```
[error]   last tree to typer: Ident(headers$1)
[error]        tree position: [...]
[error]             tree tpe: scala.scalanative.native.Ptr
[error]               symbol: value headers$1
[error]    symbol definition: headers$1: scala.scalanative.native.Ptr (a TermSymbol)
```

**Comment by tindzk** [comment]
This looks like a related bug:

```scala
def f(g: () => Unit): Unit = CFunctionPtr.fromFunction0(g)
```

It fails with this exception:

```
scala.MatchError: scala.scalanative.native.CFunctionPtr.fromFunction0(g) (of class scala.reflect.internal.Trees$Apply)
        at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genFunPtrOp(NirCodeGen.scala:1614)
        at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genPrimitiveOp(NirCodeGen.scala:1091)
        at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genApply(NirCodeGen.scala:1033)
        at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase.genExpr(NirCodeGen.scala:521)
        at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$31.apply(NirCodeGen.scala:868)
        at scala.scalanative.nscplugin.NirCodeGen$NirCodePhase$$anonfun$31.apply(NirCodeGen.scala:868)
        at scala.scalanative.nir.Focus$$anonfun$sequenced$1.apply(Focus.scala:91)
        at scala.scalanative.nir.Focus$$anonfun$sequenced$1.apply(Focus.scala:90)
        [...]
```

**Comment by densh** [comment]
Thanks for the bug report!

**Comment by matil019** [comment]
I'm beginning to think that this issue is an *unsupported feature* rather than a bug.

In C, "closures with captures" are implemented as functions with `void` pointer parameters. C functions themselves can't contain any runtime data (they are just addresses to blocks of code.)
I think it is impossible to convert a closure into a C function while retaining its original signature.

**Comment by matil019** [comment]
@densh I have been experimenting Haskell and I stumbled upon `"wrapper"`, one of the FFI features of GHC. GHC can wrap a Haskell function (with or without closure) into a C function.

The documentation is here, on "3 Function pointers": https://wiki.haskell.org/Foreign_Function_Interface

Here is a minimal example:

```c
// cee.c
void twice(void (*callb)(int)) {
    callb(1);
    callb(2);
}
```

```hs
-- hskl.hs
{-# LANGUAGE ForeignFunctionInterface #-}

import Foreign.Ptr (FunPtr, freeHaskellFunPtr)
import System.IO (hFlush, stdout)

-- A binding to the C function
foreign import ccall "twice" c_twice :: FunPtr (Int -> IO ()) -> IO ()
-- Takes a Haskell function and returns a C function pointer wrapped in IO
foreign import ccall "wrapper" mkCallb :: (Int -> IO ()) -> IO (FunPtr (Int -> IO ()))

twice :: (Int -> IO ()) -> IO ()
twice callb = do
  c_callb <- mkCallb callb
  c_twice c_callb
  freeHaskellFunPtr c_callb

main = do
  putStr "enter something: "
  hFlush stdout
  s <- getLine
  twice (\i -> putStrLn $ "<" ++ s ++ "> twice: " ++ show i)
```

compile:

```sh
ghc hskl.hs cee.c
```

example I/O:

```
enter something: foo
<foo> twice: 1
<foo> twice: 2
```

We can see a value outside callback (`"foo"` in `s` in this example) is captured.

I haven't see the implementation of this "wrapper", but it implies that a call to `mkCallb` creates a storage which stores closed over values, and `freeHaskellFunPtr` destroys it. Do you think it's possible to implement a similar mechanism in scala-native?

Also, #910 looks like a duplicate of this issue.

**Comment by densh** [comment]
Fixed in #1426
